coreax/util.py

```diff
@@ -282,19 +308,100 @@ def jit_test(
     def _fn(*args, **kwargs):
         return fn(*args, **kwargs)
 
-    assert hash(_fn) != hash(fn), "Cannot guarantee recompilation of `fn`."
+    if check_hash:
+        assert hash(_fn) != hash(fn), "Cannot guarantee recompilation of `fn`."
 
     start_time = time.time()
     block_until_ready(_fn(*fn_args, **fn_kwargs))
     end_time = time.time()
     pre_delta = end_time - start_time
+
     start_time = time.time()
     block_until_ready(_fn(*fn_args, **fn_kwargs))
     end_time = time.time()
     post_delta = end_time - start_time
+
     return pre_delta, post_delta
 
 
+def _format_number(num: float) -> str:
+    """Standardise the format of the input number."""
+    if num == 0:
+        return "0 s"
+    order = log10(num)
+    if order >= 2:  # noqa: PLR2004
+        scaled_num = num / 60
+        order = "mins"
+    elif 0 <= order < 2:  # noqa: PLR2004
+        scaled_num = num
+        order = "s"
+    elif -3 <= order < 0:  # noqa: PLR2004
+        scaled_num = 1e3 * num
+        order = "ms"
+    elif -6 <= order < -3:  # noqa: PLR2004
+        scaled_num = 1e6 * num
+        order = "\u03bcs"
+    elif -9 <= order < -6:  # noqa: PLR2004
+        scaled_num = 1e9 * num
+        order = "ns"
+    else:
+        scaled_num = 1e12 * num
+        order = "ps"
+
+    return f"{round(scaled_num, 2)} {order}"
+
+
+def speed_comparison_test(
+    function_setups: Sequence[JITCompilableFunction],
+    num_runs: int = 10,
+    log_results: bool = False,
+    check_hash: bool = False,
+) -> tuple[list[tuple[Array, Array]], dict[str, float]]:
+    """
+    Compare compilation time and runtime of a list of JIT-able functions.
+
+    :param function_setups: Sequence of instances of :class:`JITCompilableFunction`
+    :param num_runs: Number of times to average function timings over
+    :log_results: If :data:`True`, the results are formatted and logged
+    :param check_hash: If :data:`True` check that the hash of the JITted functions are
+        different to the supplied functions.
+    :return: Mean and standard deviation of compilation time and runtime for each
+        function as a list of tuples, and a dictionary of raw timings
+    """
+    num_functions = len(function_setups)
+    timings_dict = {}
+    results = []
+    for i in range(num_functions):
+        timings = jnp.zeros((num_runs, 2))
+        for j in range(num_runs):
+            timings = timings.at[j, :].set(
+                jit_test(*function_setups[i], check_hash=check_hash)
+            )
+        # Compute the time just spent on compilation
+        timings_dict[f"timings_{i}"] = timings.at[:, 0].set(
+            timings[:, 0] - timings[:, 1]
+        )
+        # Compute summary statistics
+        results.append((timings.mean(axis=0), timings.std(axis=0)))
```

Comment:
If taking my suggestion of ditching the `k` loop, save mean and SD to explicit variables. Then `results.append(mean, sd)` and logging becomes neater with e.g. `mean[0].item()`.

Reply:
Needs to be `mean[0].item()` not `mean[0][0].item()` etc.